From 4677640cba5aed8a0cf6ecffb86a34f85382a200 Mon Sep 17 00:00:00 2001
From: Richard Zhu <hongxing.zhu@nxp.com>
Date: Mon, 17 Oct 2016 15:13:56 +0800
Subject: [PATCH 1239/5242] MLK-13387-4 ARM: imx: gpcv2: correct pcie phy reg
 notifier

commit  8660b669d3dca7420514762ee31dcae1e39f1ce4 from
https://source.codeaurora.org/external/imx/linux-imx.git

1.8v of imx7d pcie phy, should be turned on after
the 1p0d(1.0v) of pcie phy is turned on.
And turned off before the 1p0d(1.0v) of pcie phy
is turned off

Signed-off-by: Richard Zhu <hongxing.zhu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/gpcv2.c |   24 ++++++++++++++++++++++--
 1 file changed, 22 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-imx/gpcv2.c b/arch/arm/mach-imx/gpcv2.c
index 04e1ecc..f2e5876 100644
--- a/arch/arm/mach-imx/gpcv2.c
+++ b/arch/arm/mach-imx/gpcv2.c
@@ -712,7 +712,12 @@ static int imx_mipi_regulator_notify(struct notifier_block *nb,
 	writel_relaxed(val | BIT(2), gpc_base + GPC_PGC_CPU_MAPPING);
 
 	switch (event) {
-	case REGULATOR_EVENT_PRE_DO_ENABLE:
+	case REGULATOR_EVENT_AFT_DO_ENABLE:
+		/*
+		 * For imx7d pcie phy, VDD18 turn on time has to wait
+		 * at least 0.1 .s after VDD10 turns on.
+		 */
+		udelay(1);
 		val = readl_relaxed(gpc_base + GPC_PU_PGC_SW_PUP_REQ);
 		writel_relaxed(val | BIT(0), gpc_base + GPC_PU_PGC_SW_PUP_REQ);
 		while (readl_relaxed(gpc_base + GPC_PU_PGC_SW_PUP_REQ) & BIT(0))
@@ -726,6 +731,11 @@ static int imx_mipi_regulator_notify(struct notifier_block *nb,
 		while (readl_relaxed(gpc_base + GPC_PU_PGC_SW_PDN_REQ) & BIT(0))
 			;
 		imx_gpcv2_set_m_core_pgc(false, GPC_PGC_MIPI_PHY);
+		/*
+		 * For imx7d pcie phy, VDD18 turn off time has to advance
+		 * at least 0.1 .s before VDD10 turns off.
+		 */
+		udelay(1);
 		break;
 	default:
 		break;
@@ -747,7 +757,12 @@ static int imx_pcie_regulator_notify(struct notifier_block *nb,
 	writel_relaxed(val | BIT(3), gpc_base + GPC_PGC_CPU_MAPPING);
 
 	switch (event) {
-	case REGULATOR_EVENT_PRE_DO_ENABLE:
+	case REGULATOR_EVENT_AFT_DO_ENABLE:
+		/*
+		 * For imx7d pcie phy, VDD18 turn on time has to wait
+		 * at least 0.1 .s after VDD10 turns on.
+		 */
+		udelay(1);
 		val = readl_relaxed(gpc_base + GPC_PU_PGC_SW_PUP_REQ);
 		writel_relaxed(val | BIT(1), gpc_base + GPC_PU_PGC_SW_PUP_REQ);
 		while (readl_relaxed(gpc_base + GPC_PU_PGC_SW_PUP_REQ) & BIT(1))
@@ -761,6 +776,11 @@ static int imx_pcie_regulator_notify(struct notifier_block *nb,
 		while (readl_relaxed(gpc_base + GPC_PU_PGC_SW_PDN_REQ) & BIT(1))
 			;
 		imx_gpcv2_set_m_core_pgc(false, GPC_PGC_PCIE_PHY);
+		/*
+		 * For imx7d pcie phy, VDD18 turn off time has to advance
+		 * at least 0.1 .s before VDD10 turns off.
+		 */
+		udelay(1);
 		break;
 	default:
 		break;
-- 
1.7.9.5

